### ISSUE TITLE
Debugger is blank when clicking on stacktrace in netmonitor first

### ISSUE THREAD

**Issue by captainbrosset** [devtools-html/debugger.html#3049]
I tested this in the latest Nightly of Firefox 55.0a1 (2017-05-30).
Here's what I did:
- Open https://www.food.ee/restaurants/vancouver/
- Open the Netmonitor panel (make sure the debugger was not init first)
- Click on the "Show filters" button in the page (top right, above the list of restaurants)
- A new request should have appears in the Netmonitor
- Click on the [JS] icon in the request row in order to open the stack trace sidebar panel
- In the sidebar panel, find the `openFilterPanel` frame and click on the link to the source code next to it
- This does not do anything, which might be related to source maps, or the netmonitor, or the debugger, not sure
- Now, click on the Debugger tab to switch to it manually instead

**the debugger is blank at this stage**

<img width="1335" alt="capture" src="https://cloud.githubusercontent.com/assets/1152698/26626195/c5d572c2-45f6-11e7-90b5-847ab657b1d6.PNG">

An error also appears in the browser console:

```
TypeError: pendingBreakpoint is undefined[Learn More] debugger.js:16544:10
TypeError: pendingBreakpoint is undefined
Stack trace:
checkPendingBreakpoint</_ref<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16544:10
step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:184
_asyncToGenerator/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:438
_asyncToGenerator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:100
checkPendingBreakpoint@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16564:13
checkPendingBreakpoints</_ref2<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16577:14
step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:184
_asyncToGenerator/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:438
_asyncToGenerator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:100
checkPendingBreakpoints@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16582:13
newSource/</_ref3<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16671:14
step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:184
step/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:362
promise callback*step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:315
_asyncToGenerator/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:438
_asyncToGenerator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:100
newSource/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16675:15
thunk/</</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:13276:46
dispatch@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:1038:19
newSources/</_ref5<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16691:16
step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:184
_asyncToGenerator/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:438
_asyncToGenerator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16637:100
newSources/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:16696:15
thunk/</</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:13276:46
bindActionCreator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:944:13
exports.onConnect</_ref<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31537:12
step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31552:184
step/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31552:362
promise callback*step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31552:315
promise callback*step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31552:315
_asyncToGenerator/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31552:438
_asyncToGenerator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31552:100
onConnect@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31548:13
onConnect</_ref<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31432:12
step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:184
_asyncToGenerator/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:438
_asyncToGenerator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:100
onConnect@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31460:13
bootstrap@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:96:15
open@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/panel.js:29:15
async*onLoad@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/framework/toolbox.js:1490:19
```

**Comment by jasonLaster** [comment]
Thanks, that bug is legit. Will fix it soon and include it in the next release this week

**Comment by codehag** [comment]
Morgen!

I am investigating this bug. Here is what I have so far:

The debugger crashes when switching from the network panel to the debugger. In fact, it also crashes if the debugger is open long enough -- it becomes unresponsive. it throws the following error:

```
"Handler function DebuggerTransport.prototype.onOutputStreamReady threw an exception: [Exception... "Component returned failure code: 0x80470007 (NS_BASE_STREAM_WOULD_BLOCK) [nsIOutputStream.write]"  nsresult: "0x80470007 (NS_BASE_STREAM_WOULD_BLOCK)"  location: "JS frame :: resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/transport/packets.js :: JSONPacket.prototype.write :: line 198"  data: no]
Stack: JSONPacket.prototype.write@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/transport/packets.js:198:17
DebuggerTransport.prototype.onOutputStreamReady<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/transport/transport.js:279:9
exports.makeInfallible/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/shared/ThreadSafeDevToolsUtils.js:101:14
Line: 198, column: 0"
```

Side note: We seem to have an issue with loading pendingBreakpoints. They are undefined if you are running the debugger with an existing profile. Its not clear why this is, because the value should be defined given that its an element being extracted from an array; Unless, that array is a list of undefined elements. this is possible. I am wondering how this happened and how to avoid it in the future. I do not think its related, but it is unexpected.

**Comment by codehag** [comment]
more notes:

the error captainbrosset posted comes in a few flavors, here is another:
```
Error: Assertion failure: js should have children
Stack trace:
assert@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:13251:12
addToTree@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:22577:7
createTree@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:22701:6
SourcesTree@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:22132:20
_constructComponentWithoutOwner@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6535:16
_constructComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6522:14
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6425:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
mountChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:13005:28
_createInitialChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:8008:27
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:7833:7
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
mountChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:13005:28
_createInitialChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:8008:27
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:7833:7
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
mountChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:13005:28
_createInitialChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:8008:27
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:7833:7
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
mountChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:13005:28
_createInitialChildren@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:8008:27
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:7833:7
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
performInitialMount@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6608:18
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:6495:16
mountComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:14744:18
mountComponentIntoNode@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:12335:16
perform@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:19035:13
batchedMountComponentIntoNode@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:12357:3
perform@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:19035:13
batchedUpdates@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:10660:7
batchedUpdates@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:16668:3
_renderNewRootComponent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:12551:5
_renderSubtreeIntoContainer@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:12632:21
render@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:12653:12
renderRoot@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:3443:6
bootstrapApp@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:32415:5
onConnect</_ref<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31454:6
step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:184
step/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:362
promise callback*step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:315
promise callback*step@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:315
_asyncToGenerator/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:438
_asyncToGenerator/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31464:100
onConnect@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:31460:13
bootstrap@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/debugger.js:96:15
open@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/debugger/new/panel.js:29:15
async*onLoad@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/framework/toolbox.js:1497:19
```

both are having something to do with loader.js, same with the exception I posted earlier

**Comment by codehag** [comment]
It looks like the prefs are being added as an immutable object, rather than an array, somewhere around here:

<img width="1147" alt="screen shot 2017-06-08 at 11 06 29 am" src="https://user-images.githubusercontent.com/26968615/26920856-91e07648-4c3a-11e7-8b3c-588ba297ee76.png">

**Comment by codehag** [comment]
The issue, as far as I can tell so far, is that we have a race condition. 

If i add too many operations before the assertion, the immutable object becomes an array. I think there is something, some inconsistancy happening higher up, but each time i try to add a debugger or a meaningful condition, the bug disappears. will keep investigating

**Comment by jasonLaster** [comment]
Added this bug for the toolbox property: https://bugzilla.mozilla.org/show_bug.cgi?id=1372840

**Comment by codehag** [comment]
this specific issue should be resolved with pr #3166. Test case:

- repeat steps mentioned by @captainbrosset 
- click around between different panels
